rust/sealevel/programs/mailbox/src/processor.rs

```diff
@@ -0,0 +1,855 @@
+use solana_program::entrypoint;
+use solana_program::{
+    account_info::next_account_info,
+    account_info::AccountInfo,
+    entrypoint::ProgramResult,
+    instruction::{AccountMeta, Instruction},
+    msg,
+    program::{invoke_signed, set_return_data},
+    program_error::ProgramError,
+    pubkey::Pubkey,
+    system_instruction,
+    sysvar::rent::Rent,
+};
+
+use crate::{
+    accounts::{Inbox, InboxAccount, Outbox, OutboxAccount},
+    error::Error,
+    instruction::{
+        InboxProcess, InboxSetDefaultModule, Init, Instruction as MailboxIxn, IsmInstruction,
+        OutboxDispatch, OutboxQuery, RecipientInstruction, MAX_MESSAGE_BODY_BYTES, VERSION,
+    },
+};
+
+#[cfg(not(feature = "no-entrypoint"))]
+entrypoint!(process_instruction);
+
+pub fn process_instruction(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    instruction_data: &[u8],
+) -> ProgramResult {
+    match MailboxIxn::from_instruction_data(instruction_data)? {
+        MailboxIxn::Init(init) => initialize(program_id, accounts, init),
+        MailboxIxn::InboxProcess(process) => inbox_process(program_id, accounts, process),
+        MailboxIxn::InboxSetDefaultModule(ism) => inbox_set_default_ism(program_id, accounts, ism),
+        MailboxIxn::OutboxDispatch(dispatch) => outbox_dispatch(program_id, accounts, dispatch),
+        MailboxIxn::OutboxGetCount(query) => outbox_get_count(program_id, accounts, query),
+        MailboxIxn::OutboxGetLatestCheckpoint(query) => {
+            outbox_get_latest_checkpoint(program_id, accounts, query)
+        }
+        MailboxIxn::OutboxGetRoot(query) => outbox_get_root(program_id, accounts, query),
+    }
+    // .map_err(|err| {
+    //     err.print::<crate::error::Error>();
+    //     err
+    // })
+}
+
+fn initialize(program_id: &Pubkey, accounts: &[AccountInfo], init: Init) -> ProgramResult {
+    // On chain create appears to use realloc which is limited to 1024 byte increments.
+    let mailbox_size = 2048;
+    let accounts_iter = &mut accounts.iter();
+
+    let system_program = next_account_info(accounts_iter)?;
+    if system_program.key != &solana_program::system_program::ID {
+        return Err(ProgramError::InvalidArgument);
+    }
+
+    let payer_account = next_account_info(accounts_iter)?;
+
+    let auth_account = next_account_info(accounts_iter)?;
+    let (auth_key, auth_bump) = Pubkey::find_program_address(
+        &[
+            b"hyperlane",
+            b"-",
+            &init.local_domain.to_le_bytes(),
+            b"-",
+            b"authority",
+        ],
+        program_id,
+    );
+    if &auth_key != auth_account.key {
+        return Err(ProgramError::InvalidArgument);
+    }
+    invoke_signed(
+        &system_instruction::create_account(
+            payer_account.key,
+            auth_account.key,
+            Rent::default().minimum_balance(0),
+            0,
+            program_id,
+        ),
+        &[
+            system_program.clone(),
```

Comment:
The conditions in the payer asserts are implied and the program should fail in other places if they are not true but it probably wouldn't hurt to add those checks.

With regards to checking the PDA owner prior to creating it, I'm assuming that is done to ensure that the account doesn't already exist and is not owned by some malicious account. I will have to dig into that a bit more however because I'm not sure about the behavior around yet to be created accounts.

Actually not sure on why `create_account()` doesn't take the `system_program` in that example - the example code I was referring to does. I will follow up on this one as well.

Reply:
Passing in `system_program` to `create_account` is unnecessary as long as the calling instruction has been passed `system_program`. The system program doesn't need itself passed in as an account